dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconMapper.java

```diff
@@ -35,71 +35,78 @@
 import org.hisp.dhis.fileresource.FileResource;
 import org.hisp.dhis.fileresource.FileResourceService;
 import org.hisp.dhis.icon.CustomIcon;
-import org.hisp.dhis.icon.DefaultIcon;
-import org.hisp.dhis.icon.Icon;
-import org.hisp.dhis.icon.IconResponse;
-import org.hisp.dhis.schema.descriptors.IconSchemaDescriptor;
-import org.hisp.dhis.user.CurrentUserUtil;
-import org.hisp.dhis.webapi.service.ContextService;
 import org.springframework.stereotype.Component;
 
 @Component
 @AllArgsConstructor
 public class IconMapper {
   private FileResourceService fileResourceService;
 
-  private ContextService contextService;
-
-  public IconResponse from(Icon icon) {
-    if (icon instanceof CustomIcon ci) {
-      return new IconResponse(
-          icon.getKey(),
-          icon.getDescription(),
-          icon.getKeywords(),
-          ci.getFileResourceUid(),
-          ci.getCreatedByUserUid(),
-          getCustomIconReference(ci.getKey()),
-          icon.getCreated(),
-          icon.getLastUpdated());
-    } else {
-      return new IconResponse(
-          icon.getKey(),
-          icon.getDescription(),
-          icon.getKeywords(),
-          getDefaultIconReference(icon.getKey()),
-          icon.getCreated(),
-          icon.getLastUpdated());
+  public CustomIcon to(CustomIconRequest customIconRequest) throws BadRequestException {
+
+    Optional<FileResource> fileResource = Optional.empty();
+
+    if (customIconRequest.getCustom()) {
+
+      if (customIconRequest.getFileResourceId() == null) {
+        throw new BadRequestException("FileResource must be provided with CustomIcon");
+      }
+
+      fileResource =
+          fileResourceService.getFileResource(customIconRequest.getFileResourceId(), CUSTOM_ICON);
+      if (fileResource.isEmpty()) {
+        throw new BadRequestException(
+            String.format(
+                "FileResource with uid %s does not exist", customIconRequest.getFileResourceId()));
+      }
     }
+
+    CustomIcon customIcon = new CustomIcon();
+    customIcon.setKey(customIconRequest.getKey());
+    customIcon.setDescription(customIconRequest.getDescription());
+    customIcon.setKeywords(customIconRequest.getKeywords());
+    customIcon.setCustom(customIconRequest.getCustom());
+    customIcon.setCode(customIconRequest.getCode());
+    customIcon.setFileResource(fileResource.orElse(null));
+
+    return customIcon;
   }
 
-  public CustomIcon to(IconDto iconDto) throws BadRequestException {
-    Optional<FileResource> fileResource =
-        fileResourceService.getFileResource(iconDto.getFileResourceUid(), CUSTOM_ICON);
-    if (fileResource.isEmpty()) {
-      throw new BadRequestException(
-          String.format("File resource with uid %s does not exist", iconDto.getFileResourceUid()));
+  public void merge(CustomIcon persisted, CustomIconRequest customIconRequest)
```

Comment:
I don't think this is a good way to implement updates. Better pass down the request object to the store layer and build SQL from it. That will avoid needing the file resource (or any other non updated) property. Service layer should be responsible for validating the update based on the request.